### PR TITLE
Store OAuth Tokens on login via Biome

### PR DESCRIPTION
**Pull request by peterschwarz**
This PR stores the oauth token values in biome, when enabled.

## Testing

Set up a local environment, with postgres as the backend.

Start a postgreSQL instance:
```bash
$ docker run --name postgres -p 5432:5432 -e POSTGRES_PASSWORD=mypassword -e POSTGRES_DB=splinter -d postgres
```
Set up the database tables:
```bash
$ cargo run --manifest-path cli/Cargo.toml -- database migrate -C postgres://postgres:mypassword@localhost:5432/splinter
```
Set the following environment variables:
```bash
export OAUTH_PROVIDER="github"
export OAUTH_CLIENT_ID="<your client id>"
export OAUTH_CLIENT_SECRET="<your client secret>"
export OAUTH_REDIRECT_URL="http://localhost:8080/oauth/callback"
```
(Or you could set these as CLI parameters for `splinterd`)

Run splinter:

```bash
$ mkdir -p /tmp/splinter # a splinter home dir for local use
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml -- cert generate --skip
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path splinterd/Cargo.toml --features auth,biome-oauth -- \
    --database postgres://postgres:mypassword@localhost:5432/splinter \
    --enable-biome \
    --tls-insecure \
    -vv
```

In your browser, navigate to [localhost:8080/oauth/login?redirect_url=http://localhost:8080/status](http://localhost:8080/oauth/login?redirect_url=http://localhost:8080/status). (Note, this will result in a 401, but will return).

After this completes, you can verify that a token has been stored in the table.  To verify the the information has been stored, run (with the password configured above `mypassword`):

```
$ docker run -it --rm --link postgres:postgres postgres psql -h postgres -U postgres
Password for user postgres:
postgres-# \c splinter
You are now connected to database "splinter" as user "postgres".
splinter=# select * from oauth_user;
 id |               user_id                | provider_user_ref |               access_token               | refresh_token | provider_id
----+--------------------------------------+-------------------+------------------------------------------+---------------+-------------
  1 | 88074e7f-18f1-47f7-ab67-65c6968127ba | your_username     | aaaaaaaaaaaaaaaaaaaaaaa00000000000000000 |               |           1
(1 row)
```

### CODE REVIEW

File: libsplinter/src/biome/oauth/store/mod.rs [hunk 1]
```diff
@@ -130,8 +130,8 @@ impl OAuthUserBuilder {
     /// Set the OAuth refresh token.
     ///
     /// This field is optional when constructing the final struct.
-    pub fn with_refresh_token(mut self, refresh_token: String) -> Self {
-        self.refresh_token = Some(refresh_token);
+    pub fn with_refresh_token(mut self, refresh_token: Option<String>) -> Self {
```
Comment (ltseeley) [inline libsplinter/src/biome/oauth/store/mod.rs:133]: What's the reason for this change?

File: libsplinter/src/rest_api/mod.rs [hunk 2]
```diff
@@ -789,6 +838,34 @@ impl RestApiBuilder {
...
+            (None, Some(_)) => {
+                return Err(RestApiServerError::MissingField(
+                    "REST API auth is not enabled as OAuth, but a save token operation has \
+                    been provided"
+                        .to_string(),
+                ));
+            }
+            (None, None) => (),
+        }
+
```
Comment (ltseeley) [inline libsplinter/src/rest_api/mod.rs:868]: This method is for starting the REST API without authentication, so this shouldn't be added.

File: libsplinter/src/biome/rest_api/auth/oauth.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+use crate::auth::{
+    oauth::{rest_api::SaveTokensOperation, UserTokens},
+    rest_api::identity::{Authorization, BearerToken, IdentityProvider},
+};
+use crate::biome::oauth::store::{OAuthProvider, OAuthUserBuilder, OAuthUserStore};
+use crate::biome::user::store::{User, UserStore};
+use crate::error::InternalError;
+
+#[derive(Clone)]
+pub struct OAuthUserStoreSaveTokensOperation {
```
Comment (ltseeley) [inline libsplinter/src/biome/rest_api/auth/oauth.rs:26]: Missing doc comments throughout this module

File: libsplinter/src/biome/oauth/store/mod.rs (hunk 1 above)
Comment (peterschwarz) [reply libsplinter/src/biome/oauth/store/mod.rs:133]: This allows the refresh token to be unset.

Comment (peterschwarz) [reply libsplinter/src/biome/oauth/store/mod.rs:133]: I could see this being in the wrong commit though, will split it out.

File: libsplinter/src/rest_api/mod.rs (hunk 2 above)
Comment (peterschwarz) [reply libsplinter/src/rest_api/mod.rs]: Fixed.

File: libsplinter/src/biome/oauth/store/mod.rs (hunk 1 above)
Comment (peterschwarz) [reply libsplinter/src/biome/oauth/store/mod.rs:133]: Split into it's own commit

File: libsplinter/src/biome/rest_api/auth/oauth.rs (hunk 3 above)
Comment (peterschwarz) [reply libsplinter/src/biome/rest_api/auth/oauth.rs:31]: Fixed.

**Review by ltseeley — DISMISSED** [review]

**Review by ltseeley — APPROVED** [review]

**Review by shannynalayna — APPROVED** [review]
Tested
